### PR TITLE
Andrewl/fix gpr posterior leading dim

### PR DESCRIPTION
**Pull request by Andrew878**
**PR type:** bugfix 

**Related issue(s)/PRs:** 

## Summary

**Proposed changes**

* The GPR posterior implementation was not working for `predict_f` calls with inputs containing leading dimensions.
* This PR addresses the problem and also amends the test so that it is covered in the future
* ...

**What alternatives have you considered?**

### Minimal working example

 

This would work:
```python
# refer to test_gpr_posterior.py
m = GPR_with_posterior(data=some_regression_data, kernel=k)
X_new = np.random.randn(5, 2)
m.posterior(cache_type).predict_f(X_new, full_cov=full_cov, full_output_cov=full_output_cov)

```
But this wouldn't
```python
# refer to test_gpr_posterior.py
m = GPR_with_posterior(data=some_regression_data, kernel=k)
X_new = np.random.randn(10, 5, 2)
m.posterior(cache_type).predict_f(X_new, full_cov=full_cov, full_output_cov=full_output_cov)

```
and would result in `InvalidArgumentError`

The PR addresses this. 
### Release notes

**Fully backwards compatible:** yes / no

**If not, why is it worth breaking backwards compatibility:**

## PR checklist

- [ ] New features: code is well-documented
  - [ ] detailed docstrings (API documentation)
  - [ ] notebook examples (usage demonstration)
- [ X] The bug case / new feature is covered by unit tests
- [ X] Code has type annotations
- [ X] Build checks
  - [X ] I ran the black+isort formatter (`make format`)
  - [X ] I locally tested that the tests pass (`make check-all`)
- [ X] Release management
  - [X ] RELEASE.md updated with entry for this change
  - [ X] New contributors: I've added myself to CONTRIBUTORS.md

### CODE REVIEW

File: gpflow/posteriors.py [hunk 1]
```diff
@@ -242,20 +242,52 @@ def _conditional_with_precompute(
...
+                tf.reshape(0, [1]),  # [M]
+                tf.reshape(K - 1, [1]),
+                tf.reshape(tf.range(1, K - 1), [K - 2]),  # leading dims (...)
+            ],
+            0,
+        )  # [N]
+        Knm = tf.transpose(Kmn, perm=perm_T)
 
+        Knn = self.kernel(Xnew, full_cov=full_cov)
         err = self.Y_data - self.mean_function(self.X_data)  # type: ignore
```
Comment (henrymoss) [inline gpflow/posteriors.py:276]: Thanks for doing this @Andrew878 . I think some of this functionality could be replaced with the ` leading_transpose` method in `gpflow.utilities.ops`

File: tests/gpflow/models/test_gpr_posterior.py [hunk 2]
```diff
@@ -21,9 +20,9 @@ def make_models(regression_data):
 
 def _get_data_for_tests():
     """Helper function to create testing data"""
-    X = np.random.randn(100, input_dim)
-    Y = np.random.randn(100, output_dim)
-    X_new = np.random.randn(100, input_dim)
+    X = np.random.randn(5, 2)
+    Y = np.random.randn(output_dim)
+    X_new = np.random.randn(10, 5, 2)
```
Comment (henrymoss) [inline tests/gpflow/models/test_gpr_posterior.py:25]: Perhaps worth testing for even more leading dims? i.e. `X_new = np.random.rand(3,10,5,2)`

File: gpflow/posteriors.py (hunk 1 above)
Comment (Andrew878) [reply gpflow/posteriors.py]: I was able to replace the `perm_T` with `leading_transpose` but I could not for the `perm` term. Happy to be shown a nice solution here (if possible)

File: tests/gpflow/models/test_gpr_posterior.py (hunk 2 above)
Comment (Andrew878) [reply tests/gpflow/models/test_gpr_posterior.py]: yes... I had to rework my solution to handle this....so good call. It now works and I've updated the tests to reflect this use case

**Comment by Andrew878** [comment]
It's not clear to me why the type checker if failing. I don't think I have used `leading_dims` any differently to other parts of the code (still looking into it)

File: tests/gpflow/models/test_gpr_posterior.py (hunk 2 above)
Comment (henrymoss) [reply tests/gpflow/models/test_gpr_posterior.py]: Great!

**Comment by Andrew878** [comment]
> It's not clear to me why the type checker if failing. I don't think I have used `leading_dims` any differently to other parts of the code (still looking into it)

I spoke to Jesper and we weren't able to understand why the mypy errors were occurring, so I have replaced `List[Union[int, type(...)]` with , `List[Any]`. While not ideal, any other input type will likely break the function (and hence flag a misspecified input). I have also made a note in the docstring about the expected type too.

**Review by frgsimpson — APPROVED** [review]
LGTM!

**Pull request merged**